pkg/kubelet/pleg/generic.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pleg
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// GenericPLEG is an extremely simple generic PLEG that relies solely on
+// periodic listing to discover container changes. It should be be used
+// as temporary replacement or container runtimes do not support a proper
+// event generator yet.
+//
+// Note that Generic PLEG assumes that a container would not be created and
+// garbage collected within one relist period. If this is not true, it might
+// miss the container completely. Relisting failure would also lead to a longer
+// window where a container could be missing. It is recommended to set the
+// relist period short and have an auxiliary, longer periodic sync in kubelet
+// the safety net.
+type GenericPLEG struct {
+	// The period for relisting.
+	relistPeriod time.Duration
+	runtime      kubecontainer.Runtime
+	eventChannel chan *PodLifecycleEvent
+	// IDs of all visible containers.
+	containers util.StringSet
+	// IDs of all running containers
+	runningContainers util.StringSet
+	// Map container ID to pod ID.
+	idMap map[string]types.UID
+}
+
+var _ PodLifecycleEventGenerator = &GenericPLEG{}
+
+func NewGenericPLEG(runtime kubecontainer.Runtime, channelCapacity int,
+	relistPeriod time.Duration) *GenericPLEG {
+	return &GenericPLEG{
+		relistPeriod:      relistPeriod,
+		runtime:           runtime,
+		eventChannel:      make(chan *PodLifecycleEvent, channelCapacity),
+		idMap:             make(map[string]types.UID),
+		containers:        util.NewStringSet(),
+		runningContainers: util.NewStringSet(),
+	}
+}
+
+// Returns a channel from which the subscriber can recieve PodLifecycleEvent
+// events.
+func (g *GenericPLEG) Watch() chan *PodLifecycleEvent {
+	return g.eventChannel
```

Comment:
With a single channel, it means only a single component can be listening to the PLEs, right? It would be useful for the ProbeManager (work in progress by me) to be able to listen for the Container{Started,Stopped} events. What if each call to Watch created a new channel, which the "main" channel cloned events into? It would add a bit more complexity around the bookkeeping of closing channels & preventing channels from backing up though.